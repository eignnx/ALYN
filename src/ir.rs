//! Intermediate Representation
#![allow(unused)]

use std::{cell::Cell, rc::Rc, sync::atomic::{AtomicUsize, Ordering}};
use internment::Intern;
use derive_more::From;

#[derive(Debug, Clone, Copy)]
pub struct Lbl(Intern<String>);

impl From<&str> for Lbl {
    fn from(name: &str) -> Self {
        Self(Intern::new(name.into()))
    }
}

pub static LBL_ID: AtomicUsize = AtomicUsize::new(0);

impl Lbl {
    pub fn fresh(base_name: &str) -> Self {
        let id = LBL_ID.fetch_add(1, Ordering::SeqCst);
        Self(Intern::new(format!("{base_name}#{id}")))
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Tmp(Intern<String>);

impl From<&str> for Tmp {
    fn from(name: &str) -> Self {
        Self(Intern::new(name.into()))
    }
}

pub static TMP_ID: AtomicUsize = AtomicUsize::new(0);

impl Tmp {
    pub fn fresh(base_name: &str) -> Self {
        let id = TMP_ID.fetch_add(1, Ordering::SeqCst);
        Self(Intern::new(format!("{base_name}#{id}")))
    }
}


#[derive(Debug, Clone, Copy)]
pub enum Binop {
    Add,
    Sub,
    Mul,
    And,
    Or,
    Xor,
}

#[derive(Debug, Clone, From)]
pub enum RVal {
    #[from]
    Int(i64),
    #[from]
    Lbl(Lbl),
    #[from]
    LVal(LVal),
    Binop(Binop, Rc<RVal>, Rc<RVal>),
    Call(Rc<RVal>, Vec<Rc<RVal>>),
    Seq(Rc<Stmt>, Rc<RVal>),
}

impl From<Tmp> for RVal {
    fn from(tmp: Tmp) -> Self {
        LVal::from(tmp).into()
    }
}

impl RVal {
    pub fn seq<const N: usize>(stmts: [Rc<Stmt>; N], result: RVal) -> Self
    {
        let mut acc = result;
        for stmt in stmts.into_iter().rev() {
            acc = RVal::Seq(stmt, Rc::new(acc));
        }
        acc
    }
}

#[derive(Debug, Clone, From)]
pub enum LVal {
    #[from]
    Tmp(Tmp),
    Mem(Rc<RVal>),
}

#[derive(Debug, Clone, Copy)]
pub enum Relop {
    Eq,
    Ne,
    Gt,
    Lt,
    GtU,
    LtU,
}

#[derive(Debug, Clone, From)]
pub enum Stmt {
    Move(LVal, RVal),
    RVal(RVal),
    Jmp(RVal, Vec<Lbl>),
    Br {
        op: Relop,
        e1: RVal,
        e2: RVal,
        if_true: Lbl,
        if_false: Lbl,
    },
    Seq(Rc<Stmt>, Rc<Stmt>),
    #[from]
    Lbl(Lbl),
}

impl Stmt {
    pub fn direct_jmp(lbl: Lbl) -> Self {
        Self::Jmp(lbl.into(), vec![lbl])
    }
}

#[test]
fn asdf() {
    // if a > b || c < d { if_true; } else { if_false; }
    let _program: Stmt = Stmt::Seq(
        Rc::new(Stmt::Br {
            op: Relop::Gt,
            e1: Tmp::from("a").into(),
            e2: Tmp::from("b").into(),
            if_true: Lbl::from("t"),
            if_false: Lbl::from("z"),
    }),
        Rc::new(Stmt::Seq(
                Rc::new(Lbl::from("z").into()),
                Rc::new(Stmt::Br {
                    op: Relop::Lt,
                    e1: Tmp::from("c").into(),
                    e2: Tmp::from("d").into(),
                    if_true: Lbl::from("t"),
                    if_false: Lbl::from("f"),
                })
        ))
    );
}
