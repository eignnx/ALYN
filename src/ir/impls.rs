use super::{Binop, IrWrap, LVal, RVal, Relop, Stmt, Unop};
use std::fmt::{Debug, Display, Formatter, Result};

impl Debug for LVal {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Param(id) => write!(f, "#{id}"),
            Self::Tmp(tmp) => write!(f, "${tmp:?}"),
            Self::Mem(rval) => write!(f, "M[{rval:?}]"),
            Self::Global(name) => write!(f, "@{name}"),
        }
    }
}

impl Debug for RVal {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Self::Byte(x) => write!(f, "{x}b"),
            Self::Nat(x) => write!(f, "{x}"),
            Self::Int(x) => write!(f, "{x:+}"),
            Self::Lbl(lbl) => write!(f, ":{lbl:?}"),
            Self::LVal(lval) => write!(f, "{lval:?}"),
            Self::Binop(op, x, y) => write!(f, "({x:?} {op} {y:?})"),
            Self::Unop(op, x) => write!(f, "({op} {x:?})"),
            Self::Call(subr, args) => {
                write!(f, "{subr:?}(")?;
                for (i, arg) in args.iter().enumerate() {
                    if i > 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{arg:?}")?;
                }
                write!(f, ")")?;
                Ok(())
            }
            Self::Seq(first, second) => {
                write!(f, "{first:?};\n")?;
                write!(f, "{second:?}")?;
                Ok(())
            }
        }
    }
}

impl Display for Binop {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Binop::Add => write!(f, "+"),
            Binop::Sub => write!(f, "-"),
            Binop::And => write!(f, "&"),
            Binop::Or => write!(f, "|"),
            Binop::Shr => write!(f, ">>"),
            Binop::Xor => write!(f, "^"),
        }
    }
}

impl Display for Relop {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Relop::Eq => write!(f, "=="),
            Relop::Ne => write!(f, "!="),
            Relop::Gt => write!(f, ">"),
            Relop::Lt => write!(f, "<"),
            Relop::Gte => write!(f, ">="),
            Relop::Lte => write!(f, "<="),
            Relop::GtU => write!(f, "u(>)"),
            Relop::LtU => write!(f, "u(<)"),
            Relop::GteU => write!(f, "u(>=)"),
            Relop::LteU => write!(f, "u(<=)"),
        }
    }
}

impl Display for Unop {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Unop::Neg => write!(f, "-"),
        }
    }
}

impl Debug for Stmt {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        match self {
            Stmt::Move(lval, rval) => write!(f, "{lval:?} <- {rval:?}"),
            Stmt::RVal(rval) => write!(f, "_ <- {rval:?}"),
            Stmt::Jmp(rval, lbls) => match &lbls[..] {
                [] => write!(f, "switch {rval:?} {{ }}"),
                [res_lbl] => match rval {
                    RVal::Lbl(cond_lbl) if cond_lbl == res_lbl => write!(f, "jmp {res_lbl:?}"),
                    other => panic!("Jmp({rval:?}, [{res_lbl:?}]) doesn't make sense!"),
                },
                lbls => {
                    write!(f, "switch {rval:?} {{\n")?;
                    for lbl in lbls {
                        write!(f, "    {lbl:?},")?;
                    }
                    write!(f, "}}")?;
                    Ok(())
                }
            },
            Stmt::Br {
                op,
                e1,
                e2,
                if_true,
                if_false,
            } => {
                write!(
                    f,
                    "if {e1:?} {op} {e2:?} then {if_true:?} else {if_false:?}"
                )
            }
            Stmt::Seq(first, second) => {
                write!(f, "{first:?};\n")?;
                write!(f, "{second:?}")?;
                Ok(())
            }
            Stmt::Lbl(lbl) => write!(f, ":{lbl:?}"),
            Stmt::Nop => write!(f, "nop"),
            Stmt::Ret(None) => write!(f, "ret"),
            Stmt::Ret(Some(x)) => write!(f, "ret {x:?}"),
        }
    }
}
