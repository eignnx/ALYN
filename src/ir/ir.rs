//! Intermediate Representation

use std::sync::atomic::{AtomicUsize, Ordering};

use derive_more::From;
use internment::Intern;

use crate::{
    names::{Lbl, Tmp},
    ty::Ty,
};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Binop {
    Add,
    Sub,
    And,
    Or,
    Shr,
    Xor,
}

#[derive(Clone, From, PartialEq)]
pub enum RVal {
    #[from]
    Byte(u8),
    #[from]
    Nat(u64),
    #[from]
    /// AKA: `Const`
    Int(i64),
    #[from]
    /// AKA: `Name`
    Lbl(Lbl),
    #[from]
    /// AKA: `Temp, Mem`
    LVal(LVal),
    Binop(Binop, Box<RVal>, Box<RVal>),
    Unop(Unop, Box<RVal>),
    BitCast(Ty, Box<RVal>),
    Call(Box<RVal>, Vec<Box<RVal>>),
    /// AKA: `ESeq`
    Seq(Box<Stmt>, Box<RVal>),
}

impl From<Tmp> for RVal {
    fn from(tmp: Tmp) -> Self {
        LVal::from(tmp).into()
    }
}

impl RVal {
    pub fn seq<const N: usize>(stmts: [Box<Stmt>; N], result: RVal) -> Self {
        let mut acc = result;
        for stmt in stmts.into_iter().rev() {
            acc = RVal::Seq(stmt, Box::new(acc));
        }
        acc
    }

    pub fn tmp(name: impl AsRef<str>) -> Self {
        Self::LVal(LVal::Tmp(name.as_ref().into()))
    }
}

#[derive(Clone, From, PartialEq)]
pub enum LVal {
    /// AKA: `Temp`
    #[from]
    Tmp(Tmp), // <-- Local
    Mem(Box<RVal>),         // <-- Deref
    Global(Intern<String>), // <-- Global
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Relop {
    Eq,
    Ne,
    Gt,
    Lt,
    Gte,
    Lte,
    GtU,
    LtU,
    GteU,
    LteU,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Unop {
    Neg,
}

#[derive(Clone, From, PartialEq)]
pub enum Stmt {
    Move(LVal, RVal),
    /// AKA: `Exp`
    RVal(RVal),
    /// AKA: `Jump`
    Jmp(RVal, Vec<Lbl>),
    /// AKA: `CJump`
    Br {
        op: Relop,
        e1: RVal,
        e2: RVal,
        if_true: Lbl,
        if_false: Lbl,
    },
    Seq(Box<Stmt>, Box<Stmt>),
    #[from]
    /// AKA: `Label`
    Lbl(Lbl),
    Nop,
    Ret(Option<RVal>),
}

impl Stmt {
    pub fn direct_jmp(lbl: impl Into<Lbl>) -> Self {
        let lbl = lbl.into();
        Self::Jmp(RVal::Lbl(lbl), vec![lbl])
    }

    pub fn seq<I>(stmts: I) -> Self
    where
        I: IntoIterator<Item = Stmt>,
        I::IntoIter: DoubleEndedIterator,
    {
        let mut iter = stmts.into_iter().rev();
        if let Some(first) = iter.next() {
            let mut acc = first;
            for stmt in iter {
                acc = Stmt::Seq(Box::new(stmt), Box::new(acc));
            }
            acc
        } else {
            Self::Nop
        }
    }
}

#[test]
fn asdf() {
    // if a > b || c < d { if_true; } else { if_false; }
    let _program: Stmt = Stmt::Seq(
        Box::new(Stmt::Br {
            op: Relop::Gt,
            e1: Tmp::from("a").into(),
            e2: Tmp::from("b").into(),
            if_true: Lbl::fresh("t"),
            if_false: Lbl::fresh("z"),
        }),
        Box::new(Stmt::Seq(
            Box::new(Lbl::fresh("z").into()),
            Box::new(Stmt::Br {
                op: Relop::Lt,
                e1: Tmp::from("c").into(),
                e2: Tmp::from("d").into(),
                if_true: Lbl::fresh("t"),
                if_false: Lbl::fresh("f"),
            }),
        )),
    );
}
