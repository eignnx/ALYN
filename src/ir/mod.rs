mod ir;
mod wrap;

pub use ir::*;
