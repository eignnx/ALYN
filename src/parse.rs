use lalrpop_util::lalrpop_mod;

lalrpop_mod!(
    #[allow(clippy::ptr_arg)]
    #[rustfmt::skip]
    grammar
);

pub use grammar::*;

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    macro_rules! test_parse {
        ($rule:ident, $src:expr) => {
            assert_debug_snapshot!(grammar::$rule::new().parse("<test>", $src));
        };
        ($rule:ident, $filename:expr, $src:expr) => {
            assert_debug_snapshot!(grammar::$rule::new().parse($filename, $src));
        };
    }

    #[test]
    fn bit_and() {
        test_parse!(RValParser, "x & y");
    }

    #[test]
    fn addr_deref_call() {
        test_parse!(RValParser, "&*x()");
    }

    #[test]
    fn pluses_and_minuses() {
        test_parse!(
            RValParser,
            "
            x - -y + z
        "
        );
    }

    #[test]
    fn operator_precedence() {
        test_parse!(
            RValParser,
            "
            x & -y + -z == 1 - 2 | 3
        "
        );
    }

    #[test]
    fn big_long_term() {
        test_parse!(
            RValParser,
            "
            *x + (127b - (+324)) - (-4) == **qwerty123 + asdf() - zxcv(1, 2)
        "
        );
    }

    #[test]
    fn type_transmutation_cast() {
        test_parse!(RValParser, "1 + 2 & bitcast{nat}(-3) & 4 + 5");
    }

    #[test]
    fn bitcast_with_nested_expr() {
        test_parse!(RValParser, "bitcast{nat}(12b + 40b)");
    }

    #[test]
    fn assignment() {
        test_parse!(StmtParser, "*lhs = 123 + *ptr;");
    }

    #[test]
    fn let_stmt() {
        test_parse!(StmtParser, "let lhs = 123 + *ptr;");
    }

    #[test]
    fn if_stmt() {
        test_parse!(StmtParser, "if 1 == 2 { lhs = 123; } else { ret asfd(); }");
    }

    #[test]
    fn simple_subr() {
        test_parse!(SubrDeclParser, "subr blah() { ret; }");
    }

    #[test]
    fn sumfac() {
        test_parse!(
            SubrDeclParser,
            "
            subr sumfac(n: nat) nat {
                ret n + sumfac(n - 1);
            }
        "
        );
    }

    #[test]
    fn main_with_args() {
        test_parse!(SubrDeclParser, "subr main(args: **byte) { ret; }");
    }

    #[test]
    fn min_subr() {
        test_parse!(
            SubrDeclParser,
            "
            subr min(a: int, b: int) int {
                if a > b {
                    ret b;
                } else {
                    ret a;
                }
            }
        "
        );
    }

    #[test]
    fn mathy_module() {
        test_parse!(
            ModuleParser,
            "mathy.alyn",
            "
            subr min(a: int, b: int) int {
                if a > b {
                    ret b;
                } else {
                    ret a;
                }
            }

            subr fib(n: nat) nat {
                if n < 2 {
                    ret 1;
                } else {
                    ret fib(n - 1) + fib(n - 2);
                }
            }
        "
        );
    }
}
