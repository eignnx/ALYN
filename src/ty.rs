use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Ty {
    Void,
    Bool,
    Byte,
    Nat,
    Int,
    Ptr(Box<Ty>),
    Subr(Vec<Ty>, Box<Ty>),
}

impl fmt::Display for Ty {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Ty::Void => write!(f, "void"),
            Ty::Bool => write!(f, "bool"),
            Ty::Byte => write!(f, "byte"),
            Ty::Nat => write!(f, "nat"),
            Ty::Int => write!(f, "int"),
            Ty::Ptr(ty) => write!(f, "*{ty}"),
            Ty::Subr(param_tys, ret_ty) => {
                write!(f, "subr(")?;
                for (i, param_ty) in param_tys.iter().enumerate() {
                    if i > 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{param_ty}")?;
                }
                write!(f, ")")?;
                if !matches!(ret_ty.as_ref(), Ty::Void) {
                    write!(f, " -> {ret_ty}")?;
                }
                Ok(())
            }
        }
    }
}
