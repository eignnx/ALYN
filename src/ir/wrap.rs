#![allow(unused)]

use crate::ir::{Binop, LVal, RVal, Relop, Stmt};
use alyn_common::names::{Lbl, Tmp};

/// AKA: `exp`
pub enum IrWrap {
    /// AKA: `Ex`
    RVal(RVal),
    /// AKA: `Nx`
    Stmt(Stmt),
    /// A closure that produces a statement from a conditional once fed both an
    /// `if_true` label and an `if_false` label.
    ///
    /// ```text
    /// // `x > y`
    /// Cond(|t, f| Br { op: Gt, e1: x, e2: y, if_true: t, if_false: f })
    ///
    /// // `if x > y { 't: (); } else { 'f: (); }`
    /// [
    ///     Br { op: Gt, e1: x, e2: y, if_true: t, if_false: f },
    ///     Lbl("t"),
    ///     Jmp("end", ["end"]),
    ///     Lbl("f"),
    ///     Lbl("end"),
    /// ]
    /// ```
    ///
    /// AKA: `Cx`
    Cond(Box<dyn FnOnce(Lbl, Lbl) -> Stmt>),
}

impl IrWrap {
    /// AKA: `unEx`
    pub fn as_expr(self) -> RVal {
        match self {
            IrWrap::RVal(rval) => rval.clone(),
            IrWrap::Stmt(stmt) => RVal::Seq(Box::new(stmt.clone()), Box::new(RVal::Int(0))),
            IrWrap::Cond(mk_stmt) => {
                let r = Tmp::fresh("r");
                let t = Lbl::fresh("t");
                let f = Lbl::fresh("f");
                RVal::seq(
                    [
                        Box::new(Stmt::Move(r.into(), 1u8.into())),
                        Box::new(mk_stmt(t, f)),
                        Box::new(f.into()),
                        Box::new(Stmt::Move(r.into(), 0u8.into())),
                        Box::new(t.into()),
                    ],
                    r.into(),
                )
            }
        }
    }

    /// AKA: `unNx`
    pub fn as_stmt(self) -> Stmt {
        match self {
            IrWrap::Stmt(stmt) => stmt.clone(),
            IrWrap::RVal(rval) => match rval {
                RVal::Seq(stmt, rest_expr) => {
                    let rest = IrWrap::RVal(*rest_expr).as_stmt();
                    Stmt::Seq(stmt, Box::new(rest))
                }
                _ => Stmt::RVal(rval.clone()),
            },
            IrWrap::Cond(mk_stmt) => {
                let after = Lbl::fresh("after_cond");
                Stmt::Seq(Box::new(mk_stmt(after, after)), Box::new(Stmt::Lbl(after)))
            }
        }
    }

    /// AKA: `unCx`
    pub fn as_cond(self) -> Box<dyn FnOnce(Lbl, Lbl) -> Stmt> {
        match self {
            IrWrap::Cond(mk_stmt) => mk_stmt,
            IrWrap::Stmt(_) => unreachable!("Statement inside condition should never occur"),
            IrWrap::RVal(RVal::Int(0)) => Box::new(|_t, f| Stmt::direct_jmp(f)),
            IrWrap::RVal(RVal::Int(_)) => Box::new(|t, _f| Stmt::direct_jmp(t)),
            IrWrap::RVal(rval) => Box::new(move |t, f| Stmt::Br {
                op: Relop::Ne,
                e1: rval.clone(),
                e2: 0u8.into(),
                if_true: t,
                if_false: f,
            }),
        }
    }

    pub fn collect_stmts_into(self, stmts: &mut Vec<Stmt>) {
        match self.as_stmt() {
            Stmt::Seq(stmt1, stmt2) => {
                IrWrap::Stmt(*stmt1).collect_stmts_into(stmts);
                IrWrap::Stmt(*stmt2).collect_stmts_into(stmts);
            }
            // In case the RVal is a Seq of stmts.
            Stmt::RVal(rval) => IrWrap::RVal(rval).collect_stmts_into(stmts),
            other => stmts.push(other),
        }
    }
}
