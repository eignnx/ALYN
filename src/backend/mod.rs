use crate::ir;

pub mod lark;

pub trait Backend<OutStream>
where
    OutStream: Extend<Self::Instruction>,
{
    type Temporary;
    type Instruction;
    fn create_backend(out: &mut OutStream) -> Self;
    fn stmt_to_asm(&mut self, stmt: ir::Stmt);
    fn expr_to_asm(&mut self, rval: ir::RVal) -> Self::Temporary;
}

pub fn run_backend<B, Tmp, Instr, Out>(out: &mut Out, stmt: ir::Stmt)
where
    B: Backend<Out, Instruction = Instr, Temporary = Tmp>,
    Out: Extend<<B as Backend<Out>>::Instruction>,
{
    let mut backend = B::create_backend(out);
    backend.stmt_to_asm(stmt)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    fn compile_to_lark(src: &str) -> Vec<lark::Instr> {
        use crate::parse;
        let mut mod_ast = parse::ModuleParser::new().parse("<test>", src).unwrap();
        use crate::ty::Ty;
        let mut tcx = crate::tcx::Tcx::new(Ty::Void);
        mod_ast.check_ty(&mut tcx);

        let mut out = vec![];
        let mut be = lark::LarkBackend::new(&mut out);
        for decl in mod_ast.decls {
            let ir = decl.value.to_ir();
            for ir_wrap in ir {
                be.stmt_to_asm(ir_wrap.as_stmt());
            }
        }
        out
    }

    #[test]
    fn basic() {
        assert_debug_snapshot!(compile_to_lark(
        "
            subr main() nat {
                ret 0;
            }
        "
        ));
    }

    #[test]
    fn binop() {
        assert_debug_snapshot!(compile_to_lark(
            "
            subr main() nat {
                let a = 24;
                ret 1 + a - 49;
            }
        "
        ));
    }

    #[test]
    fn simple_if() {
        assert_debug_snapshot!(compile_to_lark(
            "
            subr main() nat {
                if 1 == 1 {
                    let x = 1;
                } else {
                    let x = 2;
                }
            }
        "
        ));
    }
    
    #[test]
    fn simple_while() {
        assert_debug_snapshot!(compile_to_lark(
            "
            subr main() nat {
                let x = 1;
                while x < 10 {
                    x = x + 1;
                }
            }
        "
        ));
    }
}
