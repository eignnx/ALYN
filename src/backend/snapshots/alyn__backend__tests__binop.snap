---
source: src/backend/mod.rs
expression: "compile_to_lark(\"\n            subr main() nat {\n                let a = 24;\n                ret 1 + a - 49;\n            }\n        \")"
---
[
    li %a, 24,
    li %li_nat.0, 1,
    subi %imm_binop_res.1, %a, 49,
    add %binop_res.2, %li_nat.0, %imm_binop_res.1,
    jr $Ra,
]
